File: src/tree/construct-quad-tree.ts
// DIFFICULTY: Hard
//
// Given a n * n matrix grid of 0's and 1's only. We want to represent grid with a Quad-Tree.
//
// Return the root of the Quad-Tree representing grid.
//
// See https://leetcode.com/problems/construct-quad-tree/
// See https://en.wikipedia.org/wiki/Quadtree
import { _Node } from './common/quad-tree';
export { construct };

// SOLUTION:
//
// To make a quad tree, recursively divide the matrix into quadrants until each quadrant is all 1's or all 0's.
function construct(_grid: number[][]): _Node | null {
  if (_grid.length === 0 || _grid[0].length === 0) {
    return null;
  }

  return constructInternal(_grid, 0, 0, _grid.length);
}

function constructInternal(grid: number[][], row: number, column: number, size: number): _Node {
  const node = new _Node();

  // If the grid segment is uniform, we can create a single quad tree node and set the value to all 1's or 0's.
  if (isUniform(grid, row, column, size)) {
    node.val = grid[row][column] === 1;
    node.isLeaf = true;
    return node;
  }

  // If the grid segment here isn't uniform, we'll need to create 4 quadrants and construct nodes out of all 4 of
  // them.
  //
  // Setting the `node.val` of this node is irrelevant as it's not a leaf.  We can just use the default value.
  const half = size / 2;
  node.isLeaf = false;
  node.topLeft = constructInternal(grid, row, column, half);
  node.topRight = constructInternal(grid, row, column + half, half);
  node.bottomLeft = constructInternal(grid, row + half, column, half);
  node.bottomRight = constructInternal(grid, row + half, column + half, half);

  return node;
}

// Checks if a quadrant of the grid starting at [row, column] is uniformly all 1's or all 0's.
function isUniform(grid: number[][], row: number, column: number, size: number) {
  const value = grid[row][column];
  for (let i = row; i < row + size; i++) {
    for (let j = column; j < column + size; j++) {
      if (grid[i][j] !== value) {
        return false;
      }
    }
  }

  return true;
}

File: src/leetcode/tree/construct_quad_tree.py
# DIFFICULTY: HARD
# ----------------
#
# Given a n * n matrix grid of 0's and 1's only. We want to represent grid with a Quad-Tree.
#
# Return the root of the Quad-Tree representing grid.
#
# See https://leetcode.com/problems/construct-quad-tree
from leetcode.tree.common.quad_tree import Node


class Solution:
    def construct(self, grid: list[list[int]]) -> Node | None:
        """
        SOLUTION
        --------

        To make a quad tree, recursively divide the matrix into quadrants until each quadrant is all 1's or all 0's.

        COMPLEXITY
        ----------

        Time complexity is O(n^2) because we are visiting each cell in the grid.

        Space complexity is O(log n) because the recursion depth is limited by the size of the grid.
        """
        if not grid or not grid[0]:
            return None

        return self.__construct(grid, 0, 0, len(grid))

    def __construct(self, grid: list[list[int]], row: int, column: int, size: int) -> Node | None:
        node = Node()

        if self.__isUniform(grid, row, column, size):
            node.val = bool(grid[row][column])
            node.isLeaf = True
            return node

        # If the grid segment here isn't uniform, we'll need to create 4 quadrants and construct nodes out of all 4
        # of them.
        #
        # Setting the `node.val` of this node is irrelevant as it's not a leaf.  We can just use the default value.
        half = size // 2
        node.isLeaf = False
        node.topLeft = self.__construct(grid, row, column, half)
        node.topRight = self.__construct(grid, row, column + half, half)
        node.bottomLeft = self.__construct(grid, row + half, column, half)
        node.bottomRight = self.__construct(grid, row + half, column + half, half)
        return node

    # Checks if a quadrant of the grid starting at [row, column] is uniformly all 1's or all 0's.
    def __isUniform(self, grid: list[list[int]], row: int, column: int, size: int) -> bool:
        value = grid[row][column]
        for i in range(row, row + size):
            for j in range(column, column + size):
                if grid[i][j] != value:
                    return False

        return True
